presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -158,4 +197,32 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
         taskContexts.add(taskContext);
         return taskContext;
     }
+
+    private ListenableFuture<?> reserveMemoryUnsafe(long bytes)
+    {
+        if (willExceedMemoryLimit(bytes)) {
+            if (revocableFuture.isDone()) {
```

Comment:
do you need `revocableFuture` at all? Even if the future is not done, you still ask `memoryPool` for memory (so there are still new memory requests even though we are in the middle of revoke). I think it might even cause deadlock in some scenarios.

Maybe you should ask `memoryPool` only when `revocableFuture` is done?